feat: add workflow for automated vote notification

Shurtu-gal opened the pull request:
**Description**

- Checks current votes and sends weekly notification to pending members.

**Related issue(s)**
Closes #1163

Shurtu-gal commented:
/u

Shurtu-gal commented:
Sample message:
![image](https://github.com/user-attachments/assets/a06109d3-c8dd-4f55-a237-d91ef8a7eb15)

Shurtu-gal commented:
/ptal

derberg commented on .github/workflows/vote-notify.yml:
```diff
@@ -0,0 +1,152 @@
+              for (const member of leftToVote) {
+                console.log(`Notifying ${member.name} about issue #${issue.number}`);
+
+                const message = `👋 Hi ${member.name},\nYour vote is required on the following issue: *#${issue.number}*.\n*Issue Details*: <${issue.html_url}|View and cast your vote here>\nYour input is crucial to our decision-making process. Please take a moment to review the issue and share your thoughts.\nThank you for your contribution! 🙏`;
```
```suggestion
                const message = `👋 Hi ${member.name},\nWe need your vote on the following topic: *${issue.title}*.\n*Issue Details*: ${issue.html_url}\nYour input is crucial to our decision-making process. Please take a moment to review the voting topic and share your thoughts.\nThank you for your contribution! 🙏`;
```

small suggestion above:
- it is not always issue
- `issue.title` I invented this one, not sure if property like this is there, so yeah please just make sure we provide the title of issue/pr
- I tried to suggest a change to show the link without alt

derberg commented on .github/workflows/vote-notify.yml:
```diff
@@ -0,0 +1,152 @@
+        id: list
+        with:
+          script: |
+            const { data: issues } = await github.rest.issues.listForRepo({
```
will it list PRs as well?

Shurtu-gal replied:
I don't think so does `gitvote` allow PR voting as well?

derberg replied:
yeah, you either do it "manually"

```
          const { data: pullRequests } = await github.pulls.list({
            owner,
            repo: repoName,
            state: 'open'
          });
          
          const prsWithLabel = pullRequests.filter(pr => 
            pr.labels.some(label => label.name === 'vote-open')
          );
```

or use search API

```
const { data } = await github.search.issuesAndPullRequests({
            q: `repo:${owner}/${repoName} is:pr is:open label:"vote-open"`,
          });
```

I personally have bad experience with search and its accuracy, so I guess manual is better

Shurtu-gal replied:
As gitvote currently doesn't allow PR voting, marking this resolved.

derberg commented on .github/workflows/vote-notify.yml:
```diff
@@ -0,0 +1,152 @@
+      - name: Fetch Current State
+        id: fetch
+        run: |
+          cd .vote_state
```
where it comes from?

Shurtu-gal replied:
This comes from https://github.com/asyncapi/community/pull/1337/files/4a7bfe17bfcd160bf24334b2983a945167d8cf9a#diff-16c6b988d81a75b25a08240004a4a16666ba07a22160206162ccba4c1249f3d2R23

A branch stores the data between workflow runs https://github.com/ash17290/asyncapi-community/tree/vote_state

derberg replied:
ok, please extend description then, for future generations :D

derberg commented on .github/workflows/vote-notify.yml:
```diff
@@ -0,0 +1,152 @@
+                  text: message,
+                }, {
+                  headers: {
+                    'Authorization': `Bearer ${{ secrets.SLACK_TOKEN }}`,
```
we use existing one, no need to do anything here, right?

derberg replied:
yeah, there basically is no secret like `SLACK_TOKEN` - we just need to somehow create a dedicated slack token for that, this is a strategy we had so far, any publishing to slack, for any reason, always has dedicated token

derberg replied:
token under new app is under dedicated secret now: `SLACK_DM_TSC` with below scopes

<img width="569" alt="Screenshot 2024-10-22 at 09 31 07" src="https://github.com/user-attachments/assets/cbb11148-8068-4dad-bc7a-8f596cd85916">

derberg replied:
tested with below

```
curl -X POST "https://slack.com/api/chat.postMessage" \
-H "Authorization: Bearer NOTSOFASTSOLDIER" \
-H "Content-Type: application/json" \
-d '{
  "channel": "U0572R8J927", 
  "text": "Hello, this is a test message from Lukasz Gornicki related to https://github.com/asyncapi/community/pull/1337!"
}'
```

did you get it @Shurtu-gal ?

Shurtu-gal replied:
Nope I did not. Sorry took so long to revolve back to this. 😢

derberg reviewed (commented):
@Shurtu-gal left few comments, thanks a lot!

I think it is ok to make first version based on slack, and in the meantime we need to discuss with TSC if they feel fine sharing emails openly in maintainers.yml

derberg commented on .github/workflows/vote-notify.yml:
```diff
@@ -0,0 +1,152 @@
+
+                // Sending Slack DM via API
+                const response = await axios.post('https://slack.com/api/chat.postMessage', {
+                  channel: member.slack,
```
you need to either set a proper error handling here, or just check it before calling API

`slack` is added manually to `MAINTAINERS.yaml` - there is no way to get the slack id in smart automated way. So there might be situation that we overlooked it and some TSC member do not have this information provided

- we need to validate if slack property is added - and also have error handling in case wrong slack id is provided and API call fails
- in both cases we need to learn about that issue and drop notification to proper channel using `SLACK_CI_FAIL_NOTIFY` secret - you will find examples of how we do it in other workflows

Shurtu-gal replied:
Sure, give me some time I will do this as well.

Shurtu-gal replied:
@derberg, sorry to take so long. I have added it.

Shurtu-gal commented:
@derberg sorry for stretching this out. I have made the requested changes:
- Added proper desc wherever necessary.
- Added proper error handling.
- Filters only those members which have a slack id now, as many people don't have currently.

derberg commented on .github/workflows/vote-notify.yml:
```diff
@@ -0,0 +1,177 @@
+
+      - name: Notify TSC Members
+        # if: steps.list.outputs.result.length > 0
+        uses: actions/github-script@v7
```
sorry but with my experience in automation maintainance, especially of workflows with long scripts

- can you put slack notification in separate function, with proper large comment, and also pretty large comment where it is executed - so if we add other notification channels - for example using nodemailer to send email
- pleas also make other functions if possible, so the main script reads like a flow of steps
- part of the code where we determine in notification should go out, should also be separated and easy to configure - like every 3 days or only 3 days before deadline (and easy to configure to 5) for example. You know what I mean

also wondering how we can test it before merging 🤔

Shurtu-gal replied:
All suggestions have been implemented.

Test runs: https://github.com/ash17290/asyncapi-community/actions/runs/13249926799

Shurtu-gal commented:
> also wondering how we can test it before merging 🤔

I was having a test bench where I was doing it. Let me bring it up again.
https://github.com/ash17290/asyncapi-community/actions/workflows/vote-notify.yml

derberg commented on .github/workflows/vote-notify.yml:
```diff
@@ -0,0 +1,130 @@
+              for (const member of leftToVote) {
+                console.log(`Notifying ${member.name} about issue #${issue.number}`);
+
+                if (!await sendSlackNotification(member, issue, `${{ secrets.SLACK_DM_TSC }}`)) {
```
I guess I need to do something with `SLACK_DM_TSC` ?

Shurtu-gal replied:
You had already made the token I believe. https://github.com/asyncapi/community/pull/1337#discussion_r1810128100

derberg replied:
oh man, I didn't know we already discussed it in the past....didn't know it is such an old topic :D 

you are right, `SLACK_DM_TSC` is already in the repo created 4m ago :D

derberg commented on .github/workflows/vote-notify-helpers/index.js:
```diff
@@ -0,0 +1,126 @@
+ * @returns {Boolean} true if Slack notification sent successfully, false otherwise
+ */
+async function sendSlackNotification(member, issue, slackToken) {
+  const message = `👋 Hi ${member.name},\nWe need your vote on the following topic: *${issue.title}*.\n*Issue Details*: ${issue.html_url}\nYour input is crucial to our decision-making process. Please take a moment to review the voting topic and share your thoughts.\nThank you for your contribution! 🙏`;
```
can we say here how many days are left until voting is closed?

derberg commented on .github/workflows/vote-notify-helpers/index.js:
```diff
@@ -0,0 +1,126 @@
+    });
+
+    // Only 👍 and 👎 are valid votes
+    const validReactions = reactions.filter(reaction => reaction.content === '+1' || reaction.content === '-1');
```
abstain 👀 is also a valid vote

derberg commented on .github/workflows/vote-notify.yml:
```diff
@@ -0,0 +1,130 @@
+
+            const { issuesToNotify, state } = filterIssues(issues, initialState, config);
+
+            const tscMembers = yaml.load(fs.readFileSync('MAINTAINERS.yaml', 'utf8'));
```
just to clarify, it is not list of TSC members but all maintainers, and you determine who is a TSC member by certainf flag set to true - so you need to apply some filtering

Shurtu-gal replied:
Yeah, would add that.

derberg commented on .github/workflows/vote-notify.yml:
```diff
@@ -0,0 +1,130 @@
+            const issues = ${{ steps.list.outputs.result }};
+            const initialState = ${{ steps.fetch.outputs.vote_status }};
+            const config = {
+              days: 7
```
we should send notifications like half way so probably 3 days before the end

Shurtu-gal replied:
Didn't understand this point. 
Currently: 
Voting remains open for 4 weeks so we send 4 notifications each one separated by a week.

Shurtu-gal replied:
Done ✔️

derberg reviewed (commented):
oh, after refactor it is so much easier to read and review 🙏🏼 great job!

derberg commented on .github/workflows/vote-notify-helpers/index.js:
```diff
@@ -0,0 +1,132 @@
+ * 
+ * @returns {Boolean} true if Slack notification sent successfully, false otherwise
+ */
+async function sendSlackNotification(member, issue, daysBeforeDeadline, slackToken) {
```
`daysBeforeDeadline`

as discussed in another comment - predicting deadline is difficult, so please change a message to tell the voter, how much time passed since the voting request was sent

ash17290 replied:
Done.

Shurtu-gal replied:
Done

derberg commented:
/rtm

derberg commented:
@all-contributors please add @Shurtu-gal for infra, review, question, ideas, maintenance
